### PR TITLE
Parallel reduce wrapping experiment

### PR DESCRIPTION
**Pull request by orazve**
Wrapping of tbb::parallel_reduce is made.

### CODE REVIEW

File: cpp/daal/src/threading/threading.cpp [hunk 1]
```diff
@@ -111,6 +111,21 @@ DAAL_EXPORT void _daal_threader_for(int n, int threads_request, const void * a,
 #endif
 }
 
+DAAL_EXPORT int _daal_parallel_reduce(size_t n, int init, const void * a, daal::loop_functype loop_func, const void * b,
```
Comment (averbukh) [inline cpp/daal/src/threading/threading.cpp:114]: What will be function names for reduction for different types?

File: cpp/daal/src/threading/threading.h [hunk 2]
```diff
@@ -43,6 +43,10 @@ typedef void (*functype2)(int i, int n, const void * a);
 typedef void * (*tls_functype)(const void * a);
 typedef void (*tls_reduce_functype)(void * p, const void * a);
 typedef void (*functype_break)(int i, bool & needBreak, const void * a);
+
+typedef int (*loop_functype)(size_t st_idx_reduce, size_t end_idx_reduce, int value_for_reduce, const void * a);
+typedef int (*reduction_functype)(int a, int b, const void * reduction);
```
Comment (averbukh) [inline cpp/daal/src/threading/threading.h:48]: What will be type names for reduction for different types?

File: cpp/oneapi/dal/detail/threading.hpp [hunk 3]
```diff
@@ -46,6 +60,37 @@ inline ONEDAL_EXPORT void threader_for(std::int32_t n,
...
+                                              std::int32_t b,
+                                              const void *reduction) {
+    const F &lambda = *static_cast<const F *>(reduction);
+    return lambda(a, b);
+}
+
+template <typename Value, typename Func, typename Reduction>
+ONEDAL_EXPORT Value
+parallel_reduce(size_t n, Value init, const Func &func, const Reduction &reduction) {
+    const void *lf = static_cast<const void *>(&func);
```
Comment (averbukh) [inline cpp/oneapi/dal/detail/threading.hpp:83]: const void * const lf = ... ?

File: cpp/oneapi/dal/detail/threading.hpp [hunk 4]
```diff
@@ -46,6 +60,37 @@ inline ONEDAL_EXPORT void threader_for(std::int32_t n,
...
+    const void *rf = static_cast<const void *>(&reduction);
```
Comment (averbukh) [inline cpp/oneapi/dal/detail/threading.hpp:84]: const void * const rf = ... ?

File: cpp/daal/src/threading/threading.h [hunk 5]
```diff
@@ -43,6 +43,10 @@ typedef void (*functype2)(int i, int n, const void * a);
 typedef void * (*tls_functype)(const void * a);
 typedef void (*tls_reduce_functype)(void * p, const void * a);
 typedef void (*functype_break)(int i, bool & needBreak, const void * a);
+
+typedef int (*loop_functype)(size_t st_idx_reduce, size_t end_idx_reduce, int value_for_reduce, const void * a);
```
Comment (averbukh) [inline cpp/daal/src/threading/threading.h:47]: start_idx_reduce (for consistency)? See loop_functype in cpp/oneapi/dal/detail/threading.hpp

File: cpp/oneapi/dal/io/detail/load_graph.hpp [hunk 6]
```diff
@@ -61,20 +61,29 @@ void convert_to_csr_impl(const edge_list<vertex_type<Graph>> &edges, Graph &g) {
...
 
-    vertex_t max_id = edges[0].first;
-    for (auto u : edges) {
-        vertex_t edge_max = std::max(u.first, u.second);
-        max_id = std::max(max_id, edge_max);
-    }
+    vertex_t max_id = parallel_reduce(
+        edges.size(),
+        -1,
+        [&](edge_size_type st_u, edge_size_type end_u, vertex_t max_for_reduce) -> vertex_t {
```
Comment (averbukh) [inline cpp/oneapi/dal/io/detail/load_graph.hpp:77]: st_u -> start_u ?

File: cpp/oneapi/dal/io/detail/load_graph.hpp [hunk 7]
```diff
@@ -61,20 +61,29 @@ void convert_to_csr_impl(const edge_list<vertex_type<Graph>> &edges, Graph &g) {
...
+            for (auto u = st_u; u != end_u; ++u) {
+                vertex_t max_id_in_edge = std::max(edges[u].first, edges[u].second);
```
Comment (averbukh) [inline cpp/oneapi/dal/io/detail/load_graph.hpp:79]: const vertex_t

**Review by averbukh — APPROVED** [review]

**Comment by orazve** [comment]
Proposal for implementing parallel_reduce wrapping is reviewed and approved. It is not merged to master because of performance improvement is not stable for all tested datasets.

**Review by KulikovNikita — APPROVED** [review]

**Pull request closed without merging**